Added Raspberry PI 4B support with up to 4 USB-to-LAN adapters

tietzedk opened the pull request:
Add support for Raspberry PI 4B with support 4 USB-to-LAN adapters.

The Raspberry PI and USB-to-LAN adapters are easy to buy in stores and thereby it easy to replace hardware if it is broken.

lancethepants commented:
@HommeOursPorc 
I've dropped the image file here.
https://tomato64.org/files/RPI4/
Use something like balenaEtcher to flash the image to sdcard and then put into the Pi.  The built in ethernet port will be the wan port and then any USB connected ethernet dongles will be the lan ports I believe, so long as the driver and any needed firmwares are present.

HommeOursPorc commented:
@lancethepants thanks I'll check it out tomorrow 🙂

tietzedk commented:
> > > ndpi - Was that change needed to make it compile? gcc-14 - Did you attempt compiling with gcc-15? I've been jumping to the latest when available, but it isn't a hard requirement
> > 
> > 
> > I think it is better to let the kerne decide the version. I have been using local version, to differentiate the different configurations I have tested, but it is not a requirement and this can be removed from the patch
> 
> I'm good with ndpi.mk and zfs-tomato.mk changes, and whichever version of gcc you want too.
> 
> > > fstab.nvram `"tmpfs /var/log tmpfs rw,relatime 0 0"` all of /var should already be in tmpfs in `board/common/post-build-fs.sh` it sets symlink there with `ln -sf tmp/var $TARGET_DIR/var` or maybe another reason for it?
> > 
> > 
> > IMHO it is better to mount it directly and avoid softlinks that can be a bit confusing. I can remove this if you prefer symlinks.
> 
> Let's revert for now, it keeps us closer to FreshTomato functionality, and fstab.nvram is really just meant for getting nvram mounted as well.

That's fine with me

> 
> > > `mount -o remount,rw,async,nobarrier ` does this help speed up extraction on upgrade? I'm a fan.
> > 
> > 
> > The rootfs is mounted with conservative options(see cmdline.txt) to avoid a corrupt filesystem at power failure, but this has the side-effect that is slow. Remounting when doing firmware upgrade is for performance and at this point there is always a risk for corrupt filesystem/partial upgrade if a power failure occur.
> 
> Sounds good to me
> 
> > > 0001-rpi4-kernel-patch.patch Did you have script that generates this? I have a separate repo for arm64 mediatek devices that puts all the openwrt patches into a single large patch. Judging by the size of your patch it doesn't look like it's from Openwrt (which is fine). Is this generated from raspberry pi's own repo? I compile many of the kernel modules as built-in instead of as modules. I haven't looked that closely at your linux.config, but modules can be ok too. There could be some Tomato64 specific features that may expect something to be compiled in.
> > 
> > 
> > I am using the upstream kernel and this patch is only device-tree modification imported from the official RPI kernel.
> 
> So nice when manufacturers upstream their stuff
> 
> > > wlconfig are you able to use wifi on this device? That would blow my mind. If so, does your device(s) take longer to come up? or why the change?
> > 
> > 
> > WIFI seems to be working but I have only connected my laptop once to test it. I have removed the parameter from flock since it not supported by the buildroot version.
> 
> I was never able to make busybox flock work properly, so I added flock from util-linux, which should also be present on your build as well. I probably should just explicitly remove flock from being built in busybox as well.
> 
> > > zfs-tomato.mk needed to help to compile?
> > > Makefile
> > 
> > 
> > Sorry, but I do not understand the question. Could you please explain a bit more.
> 
> Just curious about the change to zfs-tomato.mk, but it's similar similar to the ndpi.mk change, so your change fine as is.
> 
> > > ```
> > > RPI4_KERNEL_VERSION=$(shell grep "BR2_LINUX_KERNEL_CUSTOM_VERSION_VALUE" tomato64/configs/mt6000_defconfig | cut -d '"' -f2)
> > > RPI4_KERNEL_PATCH=${HOME}/buildroot-src/rpi4-kernel/00001-openwrt-rpi4-kernel-${RPI4_KERNEL_VERSION}.patch
> > > ```
> > > 
> > > 
> > >     
> > >       
> > >     
> > > 
> > >       
> > >     
> > > 
> > >     
> > >   
> > > Not seeing that they are being used from what I can tell, placeholder? Also it references mt6000 config. For mediatek is used for downloading the generated kernel patch.
> > 
> > 
> > Sorry. This is a leftover from when I was not using the upstream kernel. I will remove that.
> 
> Great! that works.
> 
> If you want to rebase and then I'll merge if you like, or I can merge as is and I can make changes, whichever you prefer.

If it is okay with you, I prefer that you just do the merge as it is and make the required changes so I can focus my time on a IPv6 patch to enable support for a danish ISP.

lancethepants commented:
Well looks like you've already got one happy customer.
https://www.linksysinfo.org/index.php?threads/tomato64-development-discussion-thread.78306/page-20#post-363951

Thanks for pull request!  Looking forward to whatever more you send our way.

HommeOursPorc commented:
Thank you :)
